### PR TITLE
scripts: edtlib: Add support for 'deprecated'

### PR DESCRIPTION
**Pull request by galak**
Add the ability to mark a property as 'deprecated' to get a warning that
it will be removed in the future.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

### CODE REVIEW

**Comment by galak** [comment]
I wonder if we should also add support in gen_defines.py to mark any associated generated defines, etc as `__DEPRECATED_MACRO`.

**Comment by galak** [comment]
@mbolivar-nordic any examples of how to handle output differences on windows due to path (`/` v `\`)?

File: dts/binding-template.yaml [hunk 1]
```diff
@@ -213,6 +214,12 @@ on-bus: <string describing bus type, e.g. "i2c">
 # Note that it only makes sense to combine 'default:' with 'required: false'.
 # Combining it with 'required: true' will raise an error.
 #
+# If a property has 'deprecated: True' is a means to convey both to the user
+# and the tooling that a property is meant to be phased out.  The tooling will
+# report a warning if the devicetree includes the property that is flaged as
```
Comment (pabigot) [inline dts/binding-template.yaml:219]: s/flaged/flagged/

File: scripts/dts/edtlib.py [hunk 2]
```diff
@@ -954,6 +954,9 @@ def _prop_val(self, name, prop_type, required, default):
         # prop_type:
         #   Property type from binding (a string like "int")
         #
+        # deprecated:
```
Comment (pabigot) [inline scripts/dts/edtlib.py:957]: Curious: is there going to be any situation where a property is `deprecated` and also `required`?

Comment (mbolivar-nordic) [reply scripts/dts/edtlib.py:957]: > Curious: is there going to be any situation where a property is `deprecated` and also `required`?

There had better not

File: scripts/dts/edtlib.py [hunk 3]
```diff
@@ -906,8 +906,8 @@ def _init_prop(self, prop_spec):
         if not prop_type:
             _err("'{}' in {} lacks 'type'".format(name, self.binding_path))
 
-        val = self._prop_val(name, prop_type, prop_spec.required,
-                             prop_spec.default)
+        val = self._prop_val(name, prop_type, prop_spec.deprecated,
```
Comment (mbolivar-nordic) [inline scripts/dts/edtlib.py:909]: Is it time to just pass `prop_spec`?

**Review by mbolivar-nordic — APPROVED** [review]

File: dts/binding-template.yaml (hunk 1 above)
Comment (galak) [reply dts/binding-template.yaml:219]: fixed.

File: scripts/dts/edtlib.py (hunk 3 above)
Comment (galak) [reply scripts/dts/edtlib.py:909]: This issue is we need a "fake" prop_sec for the other caller of `_prop_val`

File: scripts/dts/edtlib.py (hunk 2 above)
Comment (galak) [reply scripts/dts/edtlib.py:957]: Added a check to error if both `deprecated` and `required` are True

**Review by mbolivar-nordic — APPROVED** [review]

File: dts/binding-template.yaml [hunk 4]
```diff
@@ -213,6 +214,12 @@ on-bus: <string describing bus type, e.g. "i2c">
 # Note that it only makes sense to combine 'default:' with 'required: false'.
 # Combining it with 'required: true' will raise an error.
 #
+# If a property has 'deprecated: True' is a means to convey both to the user
```
Comment (pabigot) [inline dts/binding-template.yaml:217]: The first sentence doesn't parse.  Maybe:
> A property with 'deprecated: True' indicates to both the user and the tooling that....

**Review by pabigot — CHANGES_REQUESTED** [review]

File: dts/binding-template.yaml (hunk 4 above)
Comment (galak) [reply dts/binding-template.yaml]: Changed.

**Review by pabigot — APPROVED** [review]

**Pull request merged**